giftless/auth/github.py

```diff
@@ -238,67 +280,226 @@ def from_dict(cls, data: Mapping[str, Any]) -> "Config":
 
+        self._api_url = self.cfg.api_url
+        self._api_headers["Authorization"] = f"Bearer {self.token}"
+        if self.cfg.api_version:
+            self._api_headers["X-GitHub-Api-Version"] = self.cfg.api_version
 
-class GithubIdentity(Identity):
-    """User identity belonging to an authentication token.
+    def __enter__(self) -> "CallContext":
+        self._session = self._exit_stack.enter_context(requests.Session())
+        self._session.headers.update(self._api_headers)
+        return self
 
-    Tracks user's permission for particular organizations/repositories.
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> Any:
+        self._session = None
+        self._exit_stack.close()
+
+    def _extract_auth(self, request: flask.Request) -> tuple[str | None, str]:
+        if request.authorization is None:
+            raise Unauthorized("Authorization required")
+
+        user = request.authorization.get("username")
+        token = request.authorization.password or request.authorization.token
+        if token is None:
+            _logger.warning(
+                f"Request to {self.org}/{self.repo} has no auth token"
+            )
+            raise Unauthorized("Authorization token required")
+        return user, token
+
+    def api_get(self, uri: str) -> dict[str, Any]:
+        if self._session is None:
+            raise RuntimeError(
+                "Calling CallContext.api_get() outside of a with block."
+            )
+        response = self._session.get(
+            f"{self._api_url}{uri}",
+            headers=self._api_headers,
+            timeout=self.cfg.api_timeout,
+        )
+        response.raise_for_status()
+        return cast(dict[str, Any], response.json())
+
+    def api_get_paginated(
+        self, uri: str, *, per_page: int = 30, list_name: str | None = None
+    ) -> Generator[dict[str, Any], None, None]:
+        if self._session is None:
+            raise RuntimeError(
+                "Calling CallContext.api_get_paginated() "
+                "outside of a with block."
+            )
+
+        per_page = min(max(per_page, 1), 100)
+        list_name = list_name or uri.rsplit("/", 1)[-1]
+        next_page = 1
+        while next_page > 0:
+            response = self._session.get(
+                f"{self._api_url}{uri}",
+                params={"per_page": per_page, "page": next_page},
+                headers=self._api_headers,
+                timeout=self.cfg.api_timeout,
+            )
+            response.raise_for_status()
+            response_json: dict[str, Any] = response.json()
+
+            yield from (item for item in response_json.get(list_name, []))
+
+            if next_url := response.links.get("next", {}).get("url"):
+                next_page = int(
+                    parse_qs(urlparse(next_url).query).get("page", ["0"])[0]
```

Comment:
A little commentary here explaining what the `["0"]` is doing and why you want the first element of that (I mean, I think I can infer it from how I think the `next_url` query should look, but it's not obvious) would be gratefully received, at least by me.